lib/edgehog_device_forwarder/http_requests/core.ex

```diff
@@ -0,0 +1,45 @@
+defmodule EdgehogDeviceForwarder.HTTPRequests.Core do
+  alias EdgehogDeviceForwarderProto.Edgehog.Device.Forwarder.Http, as: HTTP
+
+  @table EdgehogDeviceForwarder.Supervisors.Caches.for(EdgehogDeviceForwarder.HTTPRequests)
+
+  @type controller :: pid()
+  @type id :: binary()
+
+  @spec forward(controller, id, HTTP.Response.t()) :: :ok
+  def forward(controller, request_id, response) do
+    action = response_action(response)
+
+    case action do
+      :respond ->
+        ConCache.dirty_delete(@table, request_id)
+
+      {:upgrade, :websocket} ->
+        ConCache.dirty_put(@table, request_id, {:upgrading, :websocket, Qex.new()})
+    end
+
+    send(controller, {action, response})
+
+    :ok
+  end
+
+  @spec response_action(HTTP.Response.t()) :: :respond | {:upgrade, :websocket}
+  defp response_action(response) do
+    %HTTP.Response{status_code: status_code, headers: headers} = response
+
+    if status_code == 101 and has_websocket_upgrade_header?(headers) do
+      {:upgrade, :websocket}
+    else
+      :respond
+    end
+  end
+
+  @spec has_websocket_upgrade_header?(map) :: boolean()
+  defp has_websocket_upgrade_header?(headers) do
+    headers
+    |> Map.get("upgrade", "")
+    |> String.downcase(:ascii)
```

Comment:
Is this needed? Phoenix should guarantee that you always receive downcased headers

Reply:
This header map is received from a device response.
We already require the device to send the header name downcased, which was not the default.
IMO downcasing the value ourselves is better.